---
title: "Untitled"
author: "Gabe Runte"
date: "5/26/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(vegan)
library(here)
library(janitor)
library(gridExtra)
library(RColorBrewer)

```

First, I will pull in all of the sequences that I want to assign taxonomy to. Workflow is as follows:
1. Pull all of the consensus sequences
2. Combine with list of sequences not assigned to a contig
3. Map these onto the total list of sequences, allowing one to then remove samples with short sequences or with <2% high quality bp reads
4. Create final list of sequences to be assigned taxonomy
```{r}
consensus = read_csv(here("data/sanger", "sjnf_all_consensus_sep13_2023.csv")) %>% 
  clean_names() %>% 
  dplyr::select(name, sequence) %>% 
  rename(sequence_list_name = name)

contig_assign = read_csv(here("data/sanger", "sjnf_all_contigs_sep13_2023.csv")) %>% 
  clean_names() %>% 
  rename(contig_sequence = sequence) %>% 
  left_join(consensus)

metadata = read_csv(here("data/sanger", "sjnf_all_long_sep13_2023.csv")) %>% 
  clean_names() %>% 
  rename(original_seq = sequence) %>% 
  select(!sequence_list_name) %>% 
  left_join(contig_assign) %>% 
  mutate(sequence = case_when(
    is.na(sequence_list_name) ~ original_seq,
    !is.na(sequence_list_name) ~ sequence
  ))%>% 
  select(!contig_sequence) %>% 
  filter(!is.na(sequence)) %>% 
  filter(post_trim_length>75)

# taxa_set = metadata %>% 
#   select(name, percent_hq, post_trim_length, sequence, sequence_list_name) 
# 
# unite_set = taxa_set %>% 
#   select(sequence, sequence_list_name) %>% 
#   filter(is.na(sequence_list_name)) %>% 
#   bind_rows(consensus) %>% 
#   filter(str_length(sequence)>50) # only taking sequences >50bp long
```



```{r, eval = FALSE}
unite.ref <- "/Users/Gabe/Desktop/phd/ghecto/unite_db/sh_general_release_s_10.05.2021/sh_general_release_dynamic_s_10.05.2021.fasta"
unique_sequences = unique(metadata$original_seq) ###changed from my original workflow
tiptaxa = dada2::assignTaxonomy(unique_sequences, unite.ref, multithread = TRUE, tryRC = TRUE) 


unite_taxa = tibble(as.data.frame(tiptaxa)) %>% 
  mutate(sequence = rownames(tiptaxa))
write_csv(unite_taxa, here("data", "all_unite_taxa_tips_sep13.csv"))
```


```{r}
unite_taxa = read_csv( here("data", "all_unite_taxa_tips_sep13.csv")) 

field_tips = read_csv(here("data", "field_tips.csv")) %>% 
  rename(sample = dna) %>% 
  mutate(replicate= case_when(
    replicate == "Control"~ 6,
    replicate == "1" ~ 1,
    replicate == "2" ~ 2,
    replicate == "3" ~ 3,
    replicate == "4" ~ 4,
    replicate == "5" ~ 5
  ) ) %>% 
  mutate(repliacate = as.numeric(replicate))
dissected_tips_raw = read_csv(here("data", "dissect_samplesheet.csv")) %>% 
  mutate(sample = as.numeric(sample)) %>% 
  mutate(row = toupper(row))

redo_mapping = read_csv(here("data/d0607_redos.csv")) %>% 
  rename(plate = sjnf_plate) %>% 
  rename(column = old_column) %>% 
  rename(row = old_well) %>% 
  left_join(dissected_tips_raw) %>% 
  select(1,5:7) %>% 
  mutate(plate = "D0607redos") %>% 
  rename(column = new_column, row = new_well)

dissected_tips = dissected_tips_raw %>% 
  bind_rows(redo_mapping)

#poorly named sequences
should_be_names = read_csv(here("data", "pcr_success_sequencing_sheet_naming.csv")) %>% 
  rename(filename = "...7")%>% 
  separate(filename, into = c("sjnf", "p", "plate", "c", "column", "r", "row"), remove = F) %>% 
  mutate(position = rep(1:96, length.out = 236)) %>% 
  mutate(plate = as.numeric(str_sub(new_plate, 7,7))) %>% 
  select(filename, plate, position)

tip_seqs = metadata %>%
  select(!sequence) %>%################### CHANGED FROM ORIGINAL WORKFLOW #####################
  rename(sequence = original_seq) %>% ################### CHANGED FROM ORIGINAL WORKFLOW #####################
  left_join(unite_taxa) %>% 
  mutate(samset = case_when(
    str_sub(name, 6,6) == "p" ~ "cores",
    str_sub(name, 6,6) == "D" ~ "branches",
    str_sub(name, 1,1) == "p" ~ "misnamed"
  ))

misnamed = tip_seqs %>% 
  filter(samset == "misnamed") %>% 
  separate(name, into = c("plate",  "position"))%>% 
  mutate(plate = as.numeric(str_sub(plate, 2,2))) %>% 
  mutate(position = as.numeric(position)) %>% 
  left_join(should_be_names) %>% 
  rename(name = filename) %>% 
  mutate(samset = "cores")

cores = tip_seqs %>% 
  filter(samset == "cores") %>% 
  bind_rows(misnamed) %>% 
  separate(name, into = c("sjnf", "p", "plate", "c", "column", "r", "row")) %>% 
  select(!sjnf) %>% select(!p) %>% select(!c) %>% select(!r) %>% 
  mutate(plate = as.numeric(plate))%>% 
  mutate(column = as.numeric(column)) %>% 
  select(!position)

branches = tip_seqs %>% 
  filter(samset == "branches")%>% 
  separate(name, into = c("sjnf", "plate",  "column", "row")) %>% 
  select(!sjnf)%>% 
  mutate(column = as.numeric(column)) %>% 
  mutate(row = case_when(
    row == 1 ~ "A",
    TRUE ~ row
  )) %>% 
  left_join(dissected_tips) %>% 
  left_join(field_tips %>% 
              select(!c(rna, time_date)))

```

```{r}
all_trees = read_csv(here("data", "SJNF_plot surveys - SJNF_plot surveys.csv")) %>% 
  filter(str_sub(plot, 1, 4) == "POTR") 

tree_meta = all_trees %>% 
  select(plot, tag) %>% 
  mutate(elevation = tolower(str_sub(plot, 5, 5)))%>% 
  mutate(site = tolower(str_sub(plot, 6, 6))) %>% 
  select(!plot) %>% 
  rename(tree= tag) %>% 
  rename(tree_id = tree)

branch_tips = read_csv(here("data", "field_tips.csv")) %>% 
  select(site, tree_id, direction, dna, replicate) %>% 
  rename(sample = dna) %>% 
  mutate(elevation = tolower(str_sub(site, 1,1)))%>% 
  mutate(site = tolower(str_sub(site, 2,2)))

core_ids = read_csv(here("data", "soil_core_treeids - Sheet1.csv")) %>%
  left_join(tree_meta %>% 
              rename(tree = tree_id), by = "tree") %>%
  rename(elevation = elevation.x) %>%
  rename(site = site.x) %>%
  mutate(elevation = case_when(
    is.na(elevation) ~ elevation.y,
    TRUE ~ elevation
  ))%>%
  mutate(site = case_when(
    is.na(site) ~ site.y,
    TRUE ~ site
  )) %>%
  select(1:5) %>%
  mutate(row = toupper(row)) %>%
  mutate(plate = as.numeric(plate))


core_morphs = read_csv(here("data", "soil_core_morphotyping - Sheet1.csv")) %>% 
  clean_names() %>% 
  mutate(plate = as.numeric(plate))%>% 
  mutate(column = as.numeric(column)) %>%
  mutate(row = toupper(row)) %>% 
  left_join(cores) %>% 
  filter(!is.na(morphotype)) %>% 
  left_join(core_ids) %>% 
  filter(!is.na(percent_hq))

branch_morphs = read_csv(here("data", "dissect_samplesheet.csv")) %>% 
  clean_names() %>% 
  mutate(row = toupper(row)) %>% 
  left_join(branches) %>% 
  filter(!is.na(percent_hq))  %>% 
  left_join(branch_tips %>% 
              select(!site) %>% 
              mutate(replicate = as.numeric(replicate))) %>% 
  filter(!is.na(Genus))


```

```{r}
total = core_morphs %>%
  mutate(plate = as.character(plate)) %>% 
  bind_rows(branch_morphs) %>% 
  clean_names()

#write_csv(total, here("data", "sjnf_total_tipseqs.csv"))
#write_csv(branch_morphs, here("data", "sjnf_rootbranch_tipseqs.csv"))

```


```{r}

doubles = core_morphs %>%
  filter(!Genus %in% branch_morphs$Genus) %>% 
  mutate(plate = as.character(plate)) %>% 
  bind_rows(branch_morphs) %>% 
  clean_names()


ggplot(total, aes(x = elevation, fill = genus))+
  geom_bar() +
  facet_wrap(~samset)

ggplot(branch_morphs, aes(x = direction, fill = Genus))+
  geom_bar() +
  facet_wrap(~tree_id)
ggplot(core_morphs, aes(x = elevation, fill = Genus))+
  geom_bar()
```
```{r}
ggplot(branch_morphs, aes(x = direction, fill = Genus))+
  geom_bar() +
  facet_wrap(~tree_id)
```



```{r}

core_morphs = core_morphs%>% 
  filter(post_trim_length >70)

core_sum = core_morphs %>% 
  clean_names() %>% 
  group_by(genus, elevation) %>% 
  summarise(n = n()) %>% 
  left_join(core_morphs%>% 
  clean_names() %>%
  group_by(elevation) %>% 
  summarise(n_elev = n())) %>% 
  mutate(pct_elev = n/n_elev) %>% 
  mutate(elevation = factor(elevation, levels=c('l', 'm', 'h'), labels = c("Low Elev.", "Mid Elev.", "High Elev."))) %>% 
  mutate(genus = str_sub(genus, 4))


core_plot=core_sum %>% 
         filter(pct_elev >0.05)

ggplot(core_plot, aes(genus, y = pct_elev, fill = genus))+
  geom_bar(stat="identity")+
  facet_grid(rows = vars(elevation))+
  geom_text(aes(label = n), vjust = 1.5, colour = "white")+ 
  scale_fill_manual(values = colorRampPalette(brewer.pal(8, "Set2"))(24), breaks = unique(core_plot$genus), name = "Taxon")+
  theme_minimal()+
  labs(
    x = "Genus",
    y = "Relative abundance",
    title = "Soil cores"
  )

ggplot(core_plot, aes(genus, y = pct_elev, fill = genus))+
  geom_bar(stat="identity")+
  facet_grid(rows = vars(elevation))+
  geom_text(aes(label = n), vjust = 1.5, colour = "white")+ 
  scale_fill_manual(values = colorRampPalette(brewer.pal(8, "Set2"))(24), breaks = unique(core_plot$genus), name = "Taxon")+
  theme_minimal()+
  labs(
    x = "Taxon",
    y = "Relative abundance"
  )+
theme(axis.text.x=element_blank())

#ggsave(here("figures", "core_genus.jpg"), width = 7, height = 5)

branch_morphs = branch_morphs%>% 
  filter(post_trim_length >70)

branch_sum = branch_morphs %>% 
  clean_names() %>% 
  group_by(genus, elevation) %>% 
  summarise(n = n()) %>% 
  left_join(branch_morphs%>% 
  clean_names() %>% 
  group_by(elevation) %>% 
  summarise(n_elev = n())) %>% 
  mutate(pct_elev = n/n_elev)%>% 
  mutate(elevation = factor(elevation, levels=c('l', 'm', 'h'), labels = c("Low Elev.", "Mid Elev.", "High Elev."))) %>% 
  mutate(genus = str_sub(genus, 4))




ggplot(branch_sum %>% 
         filter(pct_elev >0.05) %>% 
         filter(!is.na(elevation)), aes(genus, y = pct_elev, fill = genus))+
  geom_bar(stat="identity")+
  facet_grid(rows = vars(elevation))+
  geom_text(aes(label = n), vjust = 1.5, colour = "white")+ 
  scale_fill_manual(values = colorRampPalette(brewer.pal(10, "Set2"))(18), breaks = unique(branch_sum$genus))+
  theme_minimal()+
  labs(
    x = "Genus",
    y = "Abundance",
    title = "Root branches"
  )

ggplot(branches  %>% 
         filter(Genus == "g__Russula" ) %>% 
         mutate(elevation = str_sub(site, 1,1)), aes(x=Species, fill = Species))+
  geom_bar()+
  facet_grid(rows = vars(elevation))

ggplot(branches  %>% 
         filter(Genus == "g__Cortinarius" ) %>% 
         mutate(elevation = str_sub(site, 1,1)), aes(x=Species, fill = Species))+
  geom_bar()+
  facet_grid(rows = vars(elevation))
```

What is the composition of the control root tips that are sequencing? This would be replicate 6 below. 
```{r}

branch_rep = branch_morphs %>% 
  clean_names() %>% 
  group_by(genus, replicate) %>% 
   summarise(n = n()) 

ggplot(branch_rep, aes(replicate, y = n, fill = genus))+
  geom_bar(stat="identity")+ 
  scale_fill_manual(values = colorRampPalette(brewer.pal(8, "Set2"))(28), breaks = unique(branch_rep$genus))+
  theme_minimal()+
  labs(
    x = "Replicate",
    y = "Abundance"
  )


ggplot(branch_morphs, aes(as.factor(column), fill = Genus))+
  geom_bar()+ 
  scale_fill_manual(values = colorRampPalette(brewer.pal(8, "Set2"))(28), breaks = unique(branch_morphs$Genus))+
  theme_minimal()+
  labs(
    x = "Replicate",
    y = "Abundance"
  )+
  facet_grid(rows = vars(plate))
```
 different root branches that we are pulling up all the same thing? If so, should see only one genus per bar.
```{r}

branch_comp = branch_morphs %>% 
  clean_names() %>% 
  mutate(tree_direction = paste(tree_id, direction, sep = ""))

ggplot(branch_comp %>% 
         filter(!is.na(genus)), aes(tree_direction, fill = genus))+
  geom_bar()+ 
  scale_fill_manual(values = colorRampPalette(brewer.pal(8, "Set2"))(28), breaks = unique(branch_comp$genus))+
  theme_minimal()+
  labs(
    x = "Tree/Direction",
    y = "Abundance"
  )+
  facet_wrap(~elevation)

ggplot(branch_comp %>% 
         filter(!is.na(genus)), aes(tree_direction, fill = genus))+
  geom_bar()+ 
  scale_fill_manual(values = colorRampPalette(brewer.pal(8, "Set2"))(28), breaks = unique(branch_comp$genus))+
  theme_minimal()+
  labs(
    x = "Tree/Direction",
    y = "Abundance"
  )

ggplot(branch_comp %>% 
         filter(!is.na(genus)), aes(tree_direction, fill = genus))+
  geom_bar(position = "fill")+ 
  scale_fill_manual(values = colorRampPalette(brewer.pal(8, "Set2"))(28), breaks = unique(branch_comp$genus))+
  theme_minimal()+
  labs(
    x = "Tree/Direction",
    y = "Abundance"
  )

ggplot(core_morphs, aes(x = elevation, fill = site))+
  geom_bar()
```

```{r}
branch_comp

core_morphs

ggplot(core_morphs, aes(x = elevation, fill = site))+
  geom_bar()

ggplot(core_morphs, aes(x = as.factor(plate), fill = site))+
  geom_bar()

ggplot(branch_comp, aes(x = as.factor(plate), fill = site))+
  geom_bar()
```

```{r}
russ_trees = branch_morphs %>% 
  filter(Genus == "g__Russula") %>% 
  select(site, tree_id, Genus) %>% 
  distinct() %>% 
  filter(!is.na(tree_id)) %>% 
  mutate(elevation = tolower(str_sub(site, 1,1)))%>% 
  mutate(site = tolower(str_sub(site, 2,2)))

russ_todo = russ_trees %>% 
  left_join(branch_tips) %>% 
  filter(!sample %in% branch_morphs$sample)

cort_trees = branch_morphs %>% 
  filter(Genus == "g__Cortinarius") %>% 
  select(site, tree_id, Genus) %>% 
  distinct() %>% 
  filter(!is.na(tree_id)) %>% 
  mutate(elevation = tolower(str_sub(site, 1,1)))%>% 
  mutate(site = tolower(str_sub(site, 2,2)))

cort_todo = cort_trees %>% 
  left_join(branch_tips) %>% 
  filter(!sample %in% branch_morphs$sample)

dissect_wp = read_csv(here("data", "remaining_wptips_fordissection_06302023.csv"))%>% 
  rename(site = plot)%>% 
  rename(sample = dna) %>% 
  mutate(site = tolower(site)) %>% 
  mutate(elevation = tolower(elevation)) %>% 
  arrange(sample) %>% 
  mutate(plate = rep(c("D06", "D07", "D08", "D09", "D10"), each = 96, length.out = length(sample)))%>% 
  mutate(column = rep(c(1:12), each = 8, length.out = length(sample)))%>% 
  mutate(row = rep(c("A", "B", "C", "D", "E", "F", "G", "H"), length.out = length(sample)))

#write_csv(dissect_wp, here("data", "remaining_wptips_labelled.csv"))


todo_taxa = bind_rows(cort_todo, russ_todo) %>% 
  select(-Genus) %>% 
  distinct() %>% 
  left_join(dissect_wp %>% 
              select(sample, plate, column, row)) %>% 
  arrange(sample) %>%
  filter(!is.na(plate)) %>% 
  distinct()

length(!unique(todo_taxa$sample))
```

```{r listing todo}
all_todo = total %>% 
  filter(genus %in% c("g__Cortinarius", "g__Russula")) %>% #two genera of interest for the project
  select(tree_id) %>% #just want to know the trees associated with them (given known sequences)
  distinct() %>% #just one copy of each tree #
  filter(!is.na(tree_id)) %>% #drop unknowns
  left_join(dissect_wp %>% #left join to the previous sample set to maintain consistency in plate assignments
              select(sample, plate, column, row, tree_id)) %>% 
  arrange(sample) %>% #for printing and sample finding
  filter(!is.na(plate)) %>% 
  filter(!sample %in% total$sample) %>%  #get rid of the samples we already have sequences for
  left_join(field_tips %>% 
              select(sample, replicate)) %>% 
  filter(replicate <6) %>% #remove the samples which were selected as control tips in the field
  select(-replicate)%>% 
  mutate(retry_column = rep(c(1:12), each = 8, length.out = length(sample)))%>% 
  mutate(retry_row = rep(c("A", "B", "C", "D", "E", "F", "G", "H"), length.out = length(sample))) %>% 
  filter(plate %in% c("D06", "D07"))
  
#write_csv(all_todo, here("data", "all_todissect_09182023.csv"))



```

Question: Do the communities sequences from soil cores adjacent to trees match the communities on root tips dissected in the field? 
```{r}

total_plotting = total %>% 
  mutate(tree = case_when(
    !is.na(tree) ~tree,
    is.na(tree) ~tree_id
  ))%>%
  group_by(tree) %>%
  filter(all(c("cores", "branches") %in% samset)) %>% #Here I am only interested in trees with both core and branch samples associated. Testing overlap. 
  ungroup() %>% 
  mutate(gensp = paste0(genus, species)) 
  
ggplot(total_plotting)+
  geom_bar(aes(x = samset, fill = sequence_list_name))+
  facet_wrap(~tree)

ggplot(total_plotting)+
  geom_bar(aes(x = samset, fill = gensp))+
  facet_wrap(~tree)

# I would say that yes, the communities to have significant overlap in species composition, though the relative abundance may differ of course.


total_russ = total %>% 
  filter(genus == "g__Russula") %>% 
  mutate(gensp = paste0(genus, species))

ggplot(total_russ, aes(x = elevation, fill = sequence_list_name))+
  geom_bar()+
  facet_wrap(~samset)
ggplot(total_russ, aes(x = elevation, fill = gensp))+
  geom_bar()+
  facet_wrap(~samset)
  
```
In Chunk 14: Listing Todo, I already selected for all trees that we have water potential data for and had a good idea that they would have either genus of interest. What we do not have information for is the trees in the dataset beyond D06/D07 that we can point to from the soil core data. Let's do that now. 
```{r}
todo_next = total %>% 
  filter(genus %in% c("g__Cortinarius", "g__Russula")) %>% #two genera of interest for the project
  select(tree_id) %>% #just want to know the trees associated with them (given known sequences)
  distinct() %>% #just one copy of each tree #
  filter(!is.na(tree_id)) %>% #drop unknowns
  left_join(dissect_wp %>% #left join to the previous sample set to maintain consistency in plate assignments
              select(sample, plate, column, row, tree_id)) %>% 
  arrange(sample) %>% #for printing and sample finding
  filter(!is.na(plate)) %>% 
  filter(!sample %in% total$sample) %>%  #get rid of the samples we already have sequences for
  left_join(field_tips %>% 
              select(sample, replicate)) %>% 
  filter(replicate <6) %>% #remove the samples which were selected as control tips in the field
  select(-replicate)%>% 
  mutate(retry_column = rep(c(1:12), each = 8, length.out = length(sample)))%>% 
  mutate(retry_row = rep(c("A", "B", "C", "D", "E", "F", "G", "H"), length.out = length(sample))) %>% 
  filter(!plate %in% c("D06", "D07"))


#Now we know which to specifically to target, what we also need to know is the ones we have no soil core sequences for at all. 
todo_nosc = dissect_wp %>% 
              select(sample, plate, column, row, tree_id) %>% 
  filter(!tree_id %in% core_morphs$tree)%>% 
  filter(!plate %in% c("D06", "D07"))

wp_trees = read_csv(here("data", "sample_sites.csv")) %>% 
  mutate(tree_long_new = paste0(elevation, plot, new_tag))%>% 
  mutate(tree_long_old = paste0(elevation, plot, old_tag))
  
wp_tags_all = c(wp_trees$tree_long_new, wp_trees$tree_long_old)


all_tips = read_csv(here("data", "field_tips.csv")) %>% 
  mutate(elevation = str_sub(site, 1,1)) %>% 
  mutate(plot = str_sub(site, 2,2))%>% 
  mutate(tree_long = paste0(elevation, plot, tree_id))%>% 
  filter(!tree_long %in% wp_tags_all)

all_tips_wp = read_csv(here("data", "field_tips.csv")) %>% 
  mutate(elevation = str_sub(site, 1,1)) %>% 
  mutate(plot = str_sub(site, 2,2))%>% 
  mutate(tree_long = paste0(elevation, plot, tree_id))%>% 
  filter(tree_long %in% wp_tags_all)

all_tips_update = all_tips_wp%>% 
           filter(dna>5100)


#adding in more potential tips? 
late_tips = field_tips %>% 
  select(sample, tree_id, direction, replicate) %>% 
  filter(sample %in% all_tips_update$dna)
  
```

 
 