---
title: "Untitled"
author: "Gabe Runte"
date: "5/26/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(vegan)
library(here)
library(janitor)
library(gridExtra)
library(RColorBrewer)
library(bioseq)
library(DECIPHER)
library(patchwork)
library(tidyverse)
```

First, I will pull in all of the sequences that I want to assign taxonomy to. Workflow is as follows:
1. Pull all of the consensus sequences
2. Combine with list of sequences not assigned to a contig
3. Map these onto the total list of sequences, allowing one to then remove samples with short sequences or with <2% high quality bp reads
4. Create final list of sequences to be assigned taxonomy
```{r, message = F}
folder = "data/sanger/assembly_dec13_2023"

consensus = read_csv(here(folder, "consensus.csv")) %>% 
  clean_names() %>% 
  dplyr::select(name, sequence) %>% 
  dplyr::rename(sequence_list_name = name)

contig_assign = read_csv(here(folder, "contigs.csv")) %>% 
  clean_names() %>% 
  rename(contig_sequence = sequence) %>% 
  left_join(consensus)

metadata = read_csv(here(folder, "all_seqs.csv")) %>% 
  clean_names() %>% 
  rename(original_seq = sequence) %>% 
  select(!sequence_list_name) %>% 
  left_join(contig_assign) %>% 
  mutate(sequence = case_when(
    is.na(sequence_list_name) ~ original_seq,
    !is.na(sequence_list_name) ~ sequence
  ))%>% 
  select(!contig_sequence) %>% 
  filter(!is.na(sequence))
```



```{r, eval = FALSE}
unite.ref <- "/Users/Gabe/Desktop/phd/unite/sh_general_release_dynamic_25.07.2023.fasta"
unique_sequences = unique(metadata$sequence) # using consensus seqs (change to 'originial_seq' for ASV)
#unique_sequences = unique(metadata$original_seq) 
tiptaxa = dada2::assignTaxonomy(unique_sequences, unite.ref, multithread = TRUE, tryRC = TRUE) 


unite_taxa = tibble(as.data.frame(tiptaxa)) %>% 
  mutate(sequence = rownames(tiptaxa))
write_csv(unite_taxa, here(folder, "all_unite_taxa.csv"))
#write_csv(unite_taxa, here(folder, "all_unite_taxa_nocontigs.csv"))

```



```{r, message=FALSE, warning=FALSE}
unite_taxa = read_csv(here(folder, "all_unite_taxa.csv")) 

field_tips = read_csv(here("data", "field_tips.csv")) %>% 
  rename(sample = dna)
dissected_tips_raw = read_csv(here("data", "dissect_samplesheet.csv")) %>% 
  mutate(sample = as.numeric(sample)) %>% 
  mutate(row = toupper(row)) 

dissect_control = read_csv(here("data", "controltip_dissectset_dec2023.csv")) %>% 
  rename(column = comb_column, row = comb_row) %>% 
  mutate(plate = "D13") %>% 
  select(plate, column, row, sample)

redo_mapping = read_csv(here("data/d0607_redos.csv")) %>% 
  rename(plate = sjnf_plate) %>% 
  rename(column = old_column) %>% 
  rename(row = old_well) %>% 
  left_join(dissected_tips_raw) %>% 
  select(1,5:7) %>% 
  mutate(plate = "D0607redos") %>% 
  rename(column = new_column, row = new_well)

dissected_tips = dissected_tips_raw %>% 
  bind_rows(redo_mapping) %>% 
  bind_rows(dissect_control)

#poorly named sequences
should_be_names = read_csv(here("data", "pcr_success_sequencing_sheet_naming.csv")) %>% 
  rename(filename = "...7")%>% 
  separate(filename, into = c("sjnf", "p", "plate", "c", "column", "r", "row"), remove = F) %>% 
  mutate(position = rep(1:96, length.out = 236)) %>% 
  mutate(plate = as.numeric(str_sub(new_plate, 7,7))) %>% 
  select(filename, plate, position)

tip_seqs = metadata %>%
  left_join(unite_taxa) %>% 
  mutate(samset = case_when(
    str_sub(name, 6,6) == "p" ~ "cores",
    str_sub(name, 6,6) == "D" ~ "branches",
    str_sub(name, 1,1) == "p" ~ "misnamed"
  ))

misnamed = tip_seqs %>% 
  filter(samset == "misnamed") %>% 
  separate(name, into = c("plate",  "position"))%>% 
  mutate(plate = as.numeric(str_sub(plate, 2,2))) %>% 
  mutate(position = as.numeric(position)) %>% 
  left_join(should_be_names) %>% 
  rename(name = filename) %>% 
  mutate(samset = "cores")

cores = tip_seqs %>% 
  filter(samset == "cores") %>% 
  bind_rows(misnamed) %>% 
  separate(name, into = c("sjnf", "p", "plate", "c", "column", "r", "row")) %>% 
  select(!sjnf) %>% select(!p) %>% select(!c) %>% select(!r) %>% 
  mutate(plate = as.numeric(plate))%>% 
  mutate(column = as.numeric(column)) %>% 
  select(!position)

branches = tip_seqs %>% 
  filter(samset == "branches")%>% 
  separate(name, into = c("sjnf", "plate",  "column", "row")) %>% 
  select(!sjnf)%>% 
  mutate(column = as.numeric(column)) %>% 
  mutate(row = case_when(
    row == 1 ~ "A",
    TRUE ~ row
  )) %>% 
  left_join(dissected_tips) %>% 
  left_join(field_tips %>% 
              select(!c(rna, time_date)))

```

```{r, message = F}
all_trees = read_csv(here("data", "SJNF_plot surveys - SJNF_plot surveys.csv")) %>% 
  filter(str_sub(plot, 1, 4) == "POTR") 

tree_meta = all_trees %>% 
  select(plot, tag) %>% 
  mutate(elevation = tolower(str_sub(plot, 5, 5)))%>% 
  mutate(site = tolower(str_sub(plot, 6, 6))) %>% 
  select(!plot) %>% 
  rename(tree= tag) %>% 
  rename(tree_id = tree)

branch_tips = read_csv(here("data", "field_tips.csv")) %>% 
  select(site, tree_id, direction, dna, replicate) %>% 
  rename(sample = dna) %>% 
  mutate(elevation = tolower(str_sub(site, 1,1)))%>% 
  mutate(site = tolower(str_sub(site, 2,2)))

core_ids = read_csv(here("data", "soil_core_treeids - Sheet1.csv")) %>%
  left_join(tree_meta %>% 
              rename(tree = tree_id), by = "tree") %>%
  rename(elevation = elevation.x) %>%
  rename(site = site.x) %>%
  mutate(elevation = case_when(
    is.na(elevation) ~ elevation.y,
    TRUE ~ elevation
  ))%>%
  mutate(site = case_when(
    is.na(site) ~ site.y,
    TRUE ~ site
  )) %>%
  select(1:5) %>%
  mutate(row = toupper(row)) %>%
  mutate(plate = as.numeric(plate))


core_morphs = read_csv(here("data", "soil_core_morphotyping - Sheet1.csv")) %>% 
  clean_names() %>% 
  mutate(plate = as.numeric(plate))%>% 
  mutate(column = as.numeric(column)) %>%
  mutate(row = toupper(row)) %>% 
  left_join(cores) %>% 
  filter(!is.na(morphotype)) %>% 
  left_join(core_ids) %>% 
  filter(!is.na(percent_hq))%>% 
  filter(!is.na(elevation))%>% 
  mutate(elevation = factor(elevation, levels=c("l","m","h")))

branch_morphs = dissected_tips %>% 
  clean_names() %>% 
  mutate(row = toupper(row)) %>% 
  left_join(branches) %>% 
  filter(!is.na(samset))  %>% 
  left_join(branch_tips %>% 
              select(!site)) %>% 
  filter(!is.na(Genus)) %>% 
  filter(!is.na(elevation))%>% 
  mutate(elevation = factor(elevation, levels=c("l","m","h")))


```

```{r}
total = core_morphs %>%
  mutate(plate = as.character(plate)) %>% 
  bind_rows(branch_morphs) %>% 
  clean_names()

russ_trees = branch_morphs %>% 
  filter(Genus == "g__Russula") %>% 
  select(site, tree_id, Genus) %>% 
  distinct() %>% 
  filter(!is.na(tree_id)) %>% 
  mutate(elevation = tolower(str_sub(site, 1,1)))%>% 
  mutate(site = tolower(str_sub(site, 2,2)))
```

```{r}
russ_samples = total %>% 
  filter(genus == "g__Russula") %>% 
  filter(samset == "branches") %>% 
  filter(!is.na(species))
```


```{r}
ggplot(branch_morphs, aes(x = elevation, fill = Genus))+
  geom_bar(position = "fill")

branch_genera = branch_morphs %>% 
  clean_names() %>% 
  group_by(genus) %>% 
  summarize(n = n()) %>% 
  mutate(pct = n/length(branch_morphs$sample))

top_genera = branch_genera$genus[branch_genera$n>30]

ggplot(branch_morphs %>% 
         filter(Genus %in% top_genera), aes(x = elevation, fill = Genus))+
  geom_bar(position = "fill")


russ_spe = ggplot(branch_morphs %>% 
         filter(Genus == "g__Russula"), aes(x = elevation, fill = Species))+
  geom_bar(position = "fill")+ labs(title = "Russula by 'species'")

russ_con = ggplot(branch_morphs %>% 
         filter(Genus == "g__Russula"), aes(x = elevation, fill = sequence_list_name))+
  geom_bar(position = "fill")+ labs(title = "Russula by contig")

russula_upslope = russ_spe+russ_con
#ggsave(plot = russula_upslope, here("figures", "russula_chasing.png"), width = 7, height = 5)



ggplot(branch_morphs %>% 
         filter(tree_id %in% russ_trees$tree_id), aes(x = direction, fill = Genus))+
  geom_bar()+facet_grid(rows = vars(elevation),cols = vars(tree_id))
```

```{r}
branch_ff = branch_morphs %>% 
  clean_names() %>% 
  mutate(taxonomy = paste(kingdom, phylum, class, order, family, genus, species, sep = ";"))

write_csv(branch_ff %>%  select(sample, taxonomy), here(folder, "branch_ff.csv"))

here(folder, "branch_ff.txt")
```

```{bash, engine.opts='-l', eval = F}

python /Users/Gabe/Desktop/phd/FUNGuild-master/Guilds_v1.1.py -otu /Users/Gabe/Desktop/phd/sjnf/data/sanger/assembly_dec13_2023/branch_ff.txt
```

```{r}
branch_guilds = read_csv(here(folder, "branch_ff.guilds.csv")) %>% 
  clean_names()
b_ecto = branch_morphs %>% 
  left_join(branch_guilds) %>%
  filter(str_detect(guild, "Ectomycorrhizal")) 
  
ggplot(b_ecto %>% 
         filter(Genus %in% top_genera), aes(x = direction, fill = Genus))+
  geom_bar()+facet_grid(rows = vars(elevation),cols = vars(tree_id))

ggplot(b_ecto %>% 
         filter(tree_id %in% russ_trees$tree_id), aes(x = direction, fill = Genus))+
  geom_bar()+facet_grid(rows = vars(elevation),cols = vars(tree_id))
ggplot(b_ecto %>% 
         filter(!tree_id %in% russ_trees$tree_id), aes(x = direction, fill = Genus))+
  geom_bar()+facet_grid(rows = vars(elevation),cols = vars(tree_id))

dim(branch_morphs)
dim(b_ecto)

```

## water potential tips
```{r}
wp_trees = read_csv(here("data", "sample_sites.csv")) %>% 
  mutate(tree_long_new = paste0(elevation, plot, new_tag))%>% 
  mutate(tree_long_old = paste0(elevation, plot, old_tag))
  
wp_tags_all = c(wp_trees$tree_long_new, wp_trees$tree_long_old)

all_tips_wp = read_csv(here("data", "field_tips.csv")) %>% 
  mutate(elevation = str_sub(site, 1,1)) %>% 
  mutate(plot = str_sub(site, 2,2))%>% 
  mutate(tree_long = paste0(elevation, plot, tree_id))%>% 
  filter(tree_long %in% wp_tags_all) %>% 
  rename(sample = dna)

wp_samples = all_tips_wp %>% 
  mutate(wp = TRUE) %>% 
  select(sample, wp)
```


## Control tip inquiry
```{r}
control_check = dissected_tips %>%
  left_join(branch_tips) %>% 
  filter(replicate == "Control") %>% 
  left_join(branch_morphs %>% 
              clean_names() %>% 
              select(sample, genus, species)) %>% 
  mutate(tree_site = paste(tree_id, site, sep = "_")) %>% 
  group_by(sample, elevation, tree_site, tree_id, site) %>%
  summarize(genus = if(any(!is.na(genus))) genus[which(!is.na(genus))[1]] else NA) %>% 
  ungroup() %>%  # this looks for any instance where there was a genus assignment and selects that. If none assign to the genus, it chooses NA. 
  mutate(water_potential = sample %in% all_tips_wp$sample)

ggplot(control_check, aes(x = tree_site, fill = genus, color= water_potential))+
  geom_bar()+facet_wrap(~elevation, scales = "free")
```

## Are there any more water potential trees to target? 
```{r}
wp_left = all_tips_wp %>% 
  filter(replicate == "Control") %>% 
  filter(!sample %in% control_check$sample)

## No, they've all been attempted already. 


ggplot(all_tips_wp%>% 
  mutate(tree_site = paste(tree_id, site, sep = "_")) %>% 
    filter(replicate == "Control"), aes(x = tree_site, fill = elevation))+
  geom_bar()+facet_wrap(~elevation, scales = "free_x")
```
1. Make a list of all of the trees I have enough control tips for.
2. See which of those trees I have enough identified tips from to make a set. 
3. Compare those sets to look at the variability within and between elevations? 
4. Fill out the sample set up to 7-8 trees per elevation band. 





```{r}
control3plus = control_check %>% 
  filter(is.na(genus)) %>% 
  group_by(tree_id, site, elevation) %>% 
  summarise(n = n()) %>% 
  filter(n>2)%>% 
  mutate(tree_site = paste(tree_id, site, sep = "_"))

control3plus %>% 
  group_by(elevation) %>% 
  summarize(n = n())

seq_branches = branch_morphs %>% 
  mutate(site = tolower(str_sub(site, 2,2))) %>% 
  mutate(tree_site = paste(tree_id, site, sep = "_"))%>% 
  filter(tree_site %in% control3plus$tree_site ) %>% 
  filter(replicate != "Control") %>% 
  filter(Genus %in% b_ecto$Genus)

ggplot(seq_branches, aes(x = tree_site, fill = Genus))+
  geom_bar()+facet_wrap(~elevation, scales = "free")
ggplot(seq_branches, aes(x = tree_site, fill = plate))+
  geom_bar()+facet_wrap(~elevation, scales = "free")
ggplot(seq_branches, aes(x = tree_site, fill = site))+
  geom_bar()+facet_wrap(~elevation, scales = "free")
```

Thinking about how to move forward on this project:
- Should I target broadly across more trees in H/L elevations? If I am aiming for 8ish trees per elevation. 
- I can probably do 6 trees per day. Aim for sending out by Monday-Tuesday? Begin extractions early next week? 


```{r}
trees = branch_morphs %>% 
  clean_names() %>% 
  filter(!is.na(genus)) %>% 
  filter(replicate != "Control") %>% 
  group_by(tree_id, site, elevation) %>% 
  summarise(n = n()) %>% 
  filter(n>2)%>% 
  mutate(site = tolower(str_sub(site, 2,2))) %>% 
  mutate(tree_site = paste(tree_id, site, sep = "_")) %>% 
  filter(tree_site %in% control3plus$tree_site) %>% 
  left_join(control3plus %>% 
              select(tree_site, n) %>% 
              rename(control = n)) %>% 
  mutate(sitelong = toupper(paste0(elevation, site)))

final_set = trees %>% 
  filter(control>3)

allsites = field_tips %>% 
  mutate(elevation = str_sub(site, 1,1))%>%
  rename(sitelong = site) %>% 
  mutate(site = str_sub(sitelong, 2,2)) 

donesites = tibble(sites = unique(allsites$sitelong)) %>% 
  mutate(done = sites %in% final_set$sitelong) %>% 
  filter(sites != "LB")

set.seed(219)
dissection_trees = field_tips %>% 
  filter(site %in% donesites$sites[donesites$done==F]) %>% 
  select(site, tree_id) %>% 
  distinct() %>% 
  group_by(site) %>% 
  sample_n(2)%>% 
  mutate(sitetree = paste0(site, tree_id))


dec_dissection = field_tips %>% 
  mutate(sitetree = paste0(site, tree_id)) %>% 
  filter(sitetree %in% dissection_trees$sitetree) %>% 
  arrange(sample) %>% 
  mutate(newplate = rep(c("D15", "D16", "D17", "D18"), each = 96, length.out = n())) %>% 
  mutate(row = rep(1:12, each = 8, length.out = n()))%>% 
  mutate(column = rep(c("A", "B", "C", "D", "E", "F", "G", "H"), length.out = n()))

write_csv(dec_dissection %>% 
            select(site, tree_id, direction, sample, newplate, column, row),
          here("data", "dissect_dec.csv"))
```


## EXTRACTIONS


## Workflow
1. Decide on tree to work on, select for that tree
  a. This is done by aiming for trees with 3 sterile control tips and 3+ cardinal directions with mycorrhized root tips with with high quality reads
  b. attempt to line up the cardinal directions of the ecto and control sets
2. Find control tip RNA numbers
3. For live tips:
  a. Group by direction
  b. Randomly sample within each direction that we have information for
  c. If we do not have all 4 directions, maximize directions then randomly sample from remaining tips? 
  
  optional:
  -only focus on trees with all 4 directions (how many do i have?)
  -only focus on tips with species-level assignments
  
  
```{r}
# how many trees with all 4 directions?
directional_live = branch_morphs %>% 
  clean_names() %>% 
  filter(!is.na(genus)) %>% 
  filter(replicate != "Control") %>% 
  group_by(tree_id, site, elevation) %>% 
  summarise(n = n(),
            directions = length(unique(direction))) %>% 
  filter(n>2)%>% 
  mutate(site = tolower(str_sub(site, 2,2))) %>% 
  mutate(tree_site = paste(tree_id, site, sep = "_")) %>% 
  filter(tree_site %in% control3plus$tree_site) %>% 
  left_join(control3plus %>% 
              select(tree_site, n) %>% 
              rename(control = n)) %>% 
  mutate(sitelong = toupper(paste0(elevation, site)))

# how many trees with 3+ tips at the species level??
species_live = branch_morphs %>% 
  clean_names() %>% 
  filter(!is.na(genus)) %>% 
  filter(replicate != "Control") %>% 
  group_by(tree_id, site, elevation) %>% 
  filter(!is.na(species)) %>% 
  summarise(n = n(),
            directions = length(unique(direction))) %>% 
  filter(n>2)%>% 
  mutate(site = tolower(str_sub(site, 2,2))) %>% 
  mutate(tree_site = paste(tree_id, site, sep = "_")) %>% 
  filter(tree_site %in% control3plus$tree_site) %>% 
  left_join(control3plus %>% 
              select(tree_site, n) %>% 
              rename(control = n)) %>% 
  mutate(sitelong = toupper(paste0(elevation, site)))

## In both cases, we lose a lot of trees. Perhaps we have do do our best and settle in other cases?
```
  
  


```{r}
#12/20/23
# chosen 1070 from MB and 2539 from HD
## Need to select the tips for extraction

set.seed(108)


ext_1070 = branch_morphs %>% 
  filter(tree_id == "1070") %>% 
  filter(!is.na(Species)) %>% 
  group_by(direction) %>% 
  sample_n(1) %>% 
  ungroup() %>% 
  mutate(rna = sample -100)%>% 
  select(tree_id, elevation, site, replicate, rna)

cont_1070 = dissected_tips %>% 
  left_join(branch_tips) %>% 
  filter(tree_id == "1070") %>% 
  filter(replicate == "Control") %>% 
  mutate(rna = sample -100)%>% 
  select(tree_id, elevation, site, replicate, rna)

ext_2539 = branch_morphs %>% 
  filter(tree_id == "2539") %>% 
  group_by(direction) %>% 
  slice(1) %>% 
  ungroup() %>% 
  mutate(rna = sample -100)%>% 
  select(tree_id, elevation, site, replicate, rna)

cont_2539 = dissected_tips %>% 
  left_join(branch_tips) %>% 
  filter(tree_id == "2539") %>% 
  filter(replicate == "Control") %>% 
  mutate(rna = sample -100)%>% 
  select(tree_id, elevation, site, replicate, rna)

```


```{r}
control_dir = control_check %>% 
  filter(is.na(genus)) %>% 
  left_join(branch_tips) %>% 
  mutate(control = "Y")


# how many trees with 3+ directions matching the control directions
dirs_set = branch_morphs %>% 
  clean_names() %>% 
  filter(!is.na(genus)) %>% 
  filter(replicate != "Control") %>% 
  mutate(site = tolower(str_sub(site, 2,2))) %>% 
  left_join(control_dir %>% 
              select(site, tree_id, direction, control)) %>% 
  filter(!is.na(control)) %>% 
  group_by(tree_id, site, elevation) %>% 
  summarise(n = n(),
            directions = length(unique(direction))) %>% 
  filter(n>2) %>% 
  mutate(tree_site = paste(tree_id, site, sep = "_")) %>% 
  filter(tree_site %in% control3plus$tree_site) %>% 
  left_join(control3plus %>% 
              select(tree_site, n) %>% 
              rename(control = n)) %>% 
  mutate(sitelong = toupper(paste0(elevation, site)))

#select tips from this set of tips/directions
# maximize diversity of directions, randomly sample from each direction
# create a set of matched control tips from those same 3 directions

```


```{r}
control_list = control_dir %>% 
              select(site, tree_id, direction, control)
ext_set = branch_morphs %>% 
  clean_names() %>% 
  filter(!is.na(genus))%>% 
  filter(replicate != "Control") %>% 
  mutate(site = tolower(str_sub(site, 2,2))) %>% 
  left_join(control_list) %>% 
  filter(!is.na(control))

ext_reps = ext_set %>% 
  group_by(tree_id, site, elevation) %>% 
  summarise(n = n(),
            directions = length(unique(direction))) %>% 
  filter(n>2) %>% 
  mutate(tree_site = paste(tree_id, site, sep = "_")) %>% 
  filter(tree_site %in% control3plus$tree_site) 

focal_trees = ext_reps
```

I want to know: 
Which trees that we have sequences for can be expanded upon. To do this, ideally the tree would have water potential data, 4 sterile control tips, live tips from all 4 directions. 

```{r}
tree4control = control3plus %>% 
  filter(n == 4)
treedone = ext_reps %>% 
  filter(directions == 4) %>% 
  filter(n > 3)

branch = branch_morphs %>% 
  left_join(wp_samples) %>%
  mutate(wp =case_when(
    wp == TRUE ~ TRUE, 
    is.na(wp) ~ FALSE)) %>% 
  mutate(tree_site = paste(tree_id, tolower(str_sub(site, 2,2)), sep = "_")) %>% 
  filter(tree_site %in% tree4control$tree_site) %>% 
  filter(!tree_site %in% treedone$tree_site)

## Which trees on that list have tips that still have not been dissected? 
field_tips %>% 
  filter(!sample %in% dissected_tips$sample) %>% 
  filter(tree_id %in% branch$tree_id )

## Which dissected tips belong to the trees in #branch but i have no sequence for? 

blanks = dissected_tips %>% 
  left_join(field_tips %>%  select(sample, tree_id, site)) %>% 
  mutate(site = tolower(str_sub(site, 2,2))) %>% 
  mutate(tree_site = paste(tree_id, site, sep = "_")) %>% 
  filter(tree_site %in% branch$tree_site) %>% 
  left_join(branch_morphs %>%  select(sample, Genus))

## Which directions do I still need to fill these out? 


```



